# Part 1

## Importing Libraries and Data

### Steps 1-3

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# creating path variable
path = r'/Users/paigepearson/Desktop/Data Analytics/Immersion Course/Achievement 4/Instacart Basket Analysis'

In [3]:
# importing customer data
customer_data = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
# importing latest cleaned data
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_updated2.pkl'))

### Step 4 - Data Wrangling

In [5]:
# checking for prices error
ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [6]:
# changing high price error from lesson to NaNs
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [7]:
# checking work
ords_prods_merge['prices'].describe()

count    3.239973e+07
mean     7.790994e+00
std      4.241809e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

In [8]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ... price_range_loc     busiest_day  busiest_days  \
0          0  ...       Mid-range  Regularly busy  Regular days   
1          1  ...       Mid-range  Regularly busy  Slowest days   
2          1  ...       Mid-range  Regularly busy  Slowest days   
3          1  ...       Mid-range      Least busy  Slowest days   
4          1  ...       Mid-range      Least busy  Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_spent spending_flag  \
0            Most orders        10  New customer      6.367797   Low spender   
1         Average orders        10  New customer      6.367797   Low spender   
2            Most orders        10  New customer      6.367797   Low spender   
3         Average orders        10  New customer      6.367797   Low spender   
4            Most orders        10  New customer      6.367797   Low spender   

  median_orders   order_frequency_flag  
0          20.5  Non-frequent customer  
1          20.5  Non-frequent customer  
2          20.5  Non-frequent customer  
3          20.5  Non-frequent customer  
4          20.5  Non-frequent customer  

[5 rows x 25 columns]

In [9]:
customer_data.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [10]:
# renaming columns from customer_data to align with the formatting of ords_prods_merge dataframe
customer_data = customer_data.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'number_of_dependants', 'fam_status' : 'marital_status'})

In [11]:
#checking column names
customer_data.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

### Step 5 - Data Quality and Consistency Checks

In [12]:
ords_prods_merge.dtypes

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_last_order     float64
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
_merge                   category
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
average_spent             float64
spending_flag              object
median_orders             float64
order_frequency_flag       object
dtype: object

In [13]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN         196                  1   
1                  7                   15.0         196                  1   
2                 12                   21.0         196                  1   
3                  7                   29.0         196                  1   
4                 15                   28.0         196                  1   

   reordered  ... price_range_loc     busiest_day  busiest_days  \
0          0  ...       Mid-range  Regularly busy  Regular days   
1          1  ...       Mid-range  Regularly busy  Slowest days   
2          1  ...       Mid-range  Regularly busy  Slowest days   
3          1  ...       Mid-range      Least busy  Slowest days   
4          1  ...       Mid-range      Least busy  Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_spent spending_flag  \
0            Most orders        10  New customer      6.367797   Low spender   
1         Average orders        10  New customer      6.367797   Low spender   
2            Most orders        10  New customer      6.367797   Low spender   
3         Average orders        10  New customer      6.367797   Low spender   
4            Most orders        10  New customer      6.367797   Low spender   

  median_orders   order_frequency_flag  
0          20.5  Non-frequent customer  
1          20.5  Non-frequent customer  
2          20.5  Non-frequent customer  
3          20.5  Non-frequent customer  
4          20.5  Non-frequent customer  

[5 rows x 25 columns]

In [14]:
# changing order_id and user_id to strings, unsure if I should also change product_id, aisle_id, and department_id to a string as well?
ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')

In [15]:
ords_prods_merge['order_id'] = ords_prods_merge['order_id'].astype('str')

In [16]:
ords_prods_merge.dtypes

order_id                   object
user_id                    object
eval_set                   object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_last_order     float64
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
_merge                   category
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
average_spent             float64
spending_flag              object
median_orders             float64
order_frequency_flag       object
dtype: object

In [17]:
customer_data.shape

(206209, 10)

In [18]:
customer_data.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
marital_status          object
income                   int64
dtype: object

In [19]:
customer_data.describe()

user_id            age  number_of_dependants         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [20]:
# user_id should be excluded from number analysis and the column registered as a string to avoid this
customer_data['user_id'] = customer_data['user_id'].astype('str')

In [21]:
# checking work
customer_data.dtypes

user_id                 object
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
marital_status          object
income                   int64
dtype: object

In [22]:
customer_data.describe()

age  number_of_dependants         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

In [23]:
# checking for missing data
customer_data.isnull().sum()

user_id                     0
first_name              11259
last_name                   0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependants        0
marital_status              0
income                      0
dtype: int64

In [24]:
# creating a subset to find the nan values
customer_nan = customer_data[customer_data['first_name'].isnull() == True]

In [25]:
customer_nan.head()

user_id first_name last_name  gender      state  age date_joined  \
53    76659        NaN   Gilbert    Male   Colorado   26    1/1/2017   
73    13738        NaN     Frost  Female  Louisiana   39    1/1/2017   
82    89996        NaN    Dawson  Female     Oregon   52    1/1/2017   
99    96166        NaN   Oconnor    Male   Oklahoma   51    1/1/2017   
105   29778        NaN    Dawson  Female       Utah   63    1/1/2017   

     number_of_dependants marital_status  income  
53                      2        married   41709  
73                      0         single   82518  
82                      3        married  117099  
99                      1        married  155673  
105                     3        married  151819

In [26]:
# checking for mixed data types
for col in customer_data.columns.tolist():
    weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customer_data[weird]) > 0:
        print (col)

first_name


In [27]:
# changing NaNs to strings so the first_name column has only one type of data
customer_data['first_name'] = customer_data['first_name'].astype('str')

In [28]:
# checking work on mixed data types
for col in customer_data.columns.tolist():
    weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customer_data[weird]) > 0:
        print (col)

Note: no more mixed data types found

In [29]:
# checking for duplicates
customer_dups = customer_data[customer_data.duplicated()]

In [30]:
customer_dups.shape

(0, 10)

In [31]:
customer_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependants, marital_status, income]
Index: []

### Step 6 - Combine your customer data with the rest of your prepared Instacart data

In [32]:
# double checking data types of the dataframes
ords_prods_merge.dtypes

order_id                   object
user_id                    object
eval_set                   object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_last_order     float64
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
_merge                   category
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
average_spent             float64
spending_flag              object
median_orders             float64
order_frequency_flag       object
dtype: object

In [33]:
customer_data.dtypes

user_id                 object
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependants     int64
marital_status          object
income                   int64
dtype: object

In [34]:
# taking out the past _merge column to prepare for merge
del ords_prods_merge['_merge']

In [35]:
# checking shape to compare before and after merge
ords_prods_merge.shape

(32404859, 24)

In [36]:
customer_data.shape

(206209, 10)

In [37]:
# merging ords_prods_merge and customer_data
df_merged = customer_data.merge(ords_prods_merge, on = 'user_id', indicator = True)

In [38]:
# checking work
df_merged.shape

(32404859, 34)

Note: columns looks correct, and the customer_data may not have made extra entries so a total of 32,404,859 might be correct in this case

In [39]:
df_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [40]:
df_merged.head()

user_id first_name last_name  gender     state  age date_joined  \
0   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   number_of_dependants marital_status  income  ...     busiest_day  \
0                     3        married  165665  ...  Regularly busy   
1                     3        married  165665  ...  Regularly busy   
2                     3        married  165665  ...  Regularly busy   
3                     3        married  165665  ...  Regularly busy   
4                     3        married  165665  ...  Regularly busy   

   busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
0  Busiest days            Most orders          8  New customer   
1  Regular days            Most orders          8  New customer   
2  Busiest days            Most orders          8  New customer   
3  Regular days            Most orders          8  New customer   
4  Slowest days            Most orders          8  New customer   

   average_spent  spending_flag  median_orders  order_frequency_flag _merge  
0       7.988889    Low spender           19.0      Regular customer   both  
1       7.988889    Low spender           19.0      Regular customer   both  
2       7.988889    Low spender           19.0      Regular customer   both  
3       7.988889    Low spender           19.0      Regular customer   both  
4       7.988889    Low spender           19.0      Regular customer   both  

[5 rows x 34 columns]

### Steps 7-9, check work then export new dataframe as a pickle file

In [41]:
# exporting new merged dataframe
df_merged.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'instacart_merged.pkl'))